REGRESSION(/276827@main): Large animated images are always decoded asynchronously

shallawa opened the pull request:
#### a3254962b06ea7d5c00a91bd11a609c1ad47a167
<pre>
REGRESSION(/276827@main): Large animated images are always decoded asynchronously
<a href="https://bugs.webkit.org/show_bug.cgi?id=278593">https://bugs.webkit.org/show_bug.cgi?id=278593</a>
<a href="https://rdar.apple.com/134597661">rdar://134597661</a>

Reviewed by Simon Fraser.

The large animated images should be decoded asynchronously if the setting
AnimatedImageAsyncDecodingEnabled is true. The only exception is the first frame
has to be decoded synchronously only for the first time.

* LayoutTests/fast/images/async-image-large-animated-disabled-expected.txt: Added.
* LayoutTests/fast/images/async-image-large-animated-disabled.html: Added.
* LayoutTests/fast/images/resources/animated-red-green-blue-400x400.gif: Added.
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::startAnimation):
(WebCore::BitmapImageSource::imageFrameDecodeAtIndexHasFinished):
(WebCore::BitmapImageSource::cacheNativeImageAtIndex):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):

Canonical link: <a href="https://commits.webkit.org/283753@main">https://commits.webkit.org/283753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a3f14656086d857869d100b4553b8afec1e799e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67240 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19872 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71283 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18373 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69358 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54417 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18168 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53888 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12344 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70307 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42846 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58178 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34422 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39518 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15571 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16733 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61484 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15912 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72985 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11199 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15248 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61371 "Found 1 new test failure: webanimations/accelerated-animation-opacity-animation-begin-time-after-scale-animation-ends.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11232 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58236 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61452 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14911 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9187 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2784 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42424 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43501 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44687 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43242 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/BitmapImageSource.cpp:
```diff
@@ -607,6 +602,11 @@ Expected<Ref<NativeImage>, DecodingStatus> BitmapImageSource::nativeImageAtIndex
 {
     startAnimation(subsamplingLevel, options);
+
+    // The first frame has to be decoded synchronously for the first time.
```
The comment should explain why.

shallawa replied:
I added a comment.

smfr commented on Source/WebCore/platform/graphics/BitmapImageSource.cpp:
```diff
@@ -607,6 +602,11 @@ Expected<Ref<NativeImage>, DecodingStatus> BitmapImageSource::nativeImageAtIndex
+
+    // The first frame has to be decoded synchronously for the first time.
+    if (options.decodingMode() == DecodingMode::Asynchronous && isAnimated() && !hasEverAnimated())
+        return nativeImageAtIndexForDrawing(currentFrameIndex(), subsamplingLevel, { DecodingMode::Synchronous, options.sizeForDrawing() });
```
Rather than repeat the call to `nativeImageAtIndexForDrawing()` inside the clause here, it would be cleaner to just adjust `options` and have a single call to `nativeImageAtIndexForDrawing()`

shallawa replied:
Done.

smfr commented on Source/WebCore/platform/graphics/BitmapImageSource.h:
```diff
@@ -82,6 +82,7 @@ class BitmapImageSource : public ImageSource {
     DecodingStatus frameDecodingStatusAtIndex(unsigned index) const final;
 
     // Testing support
+    URL sourceURL() const;
```
This says "testing support" but we always call `sourceURL` for async decoding (but only to log it)

shallawa replied:
This new function is removed.

smfr commented on Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp:
```diff
@@ -86,16 +86,16 @@ void ImageFrameWorkQueue::start()
                 // The WorkQueue may have been recreated before the frame was decoded.
                 if (protectedWorkQueue.ptr() != protectedThis->m_workQueue || protectedSource.ptr() != protectedThis->m_source.get()) {
-                    LOG(Images, "ImageFrameWorkQueue::%s - %p - url: %s. WorkQueue was recreated at index = %d.", __FUNCTION__, protectedThis.ptr(), protectedSource->sourceUTF8(), request.index);
+                    LOG(Images, "ImageFrameWorkQueue::%s - %p - url: %s. WorkQueue was recreated at index = %d.", __FUNCTION__, protectedThis.ptr(), sourceURL.utf8().data(), request.index);
```
Do we really need to copy sourceURL twice, just for logging? Can't we just get it from protectedSource again, or might it have changed?

shallawa replied:
Yes it is true that we can get the sourceURL from protectedSource() as long as we make the call on the main thread. I was trying to log the decoding time on the decoding thread. So when adding the log message, a crash happened because of accessing the sourceURL String from another thread. Because currently there is no log message on the decoding thread, it is better to save the String copy time. So I removed this new copyIsolated() code.

smfr commented on Source/WebCore/platform/graphics/BitmapImageSource.cpp:
```diff
@@ -609,7 +604,15 @@ Expected<Ref<NativeImage>, DecodingStatus> BitmapImageSource::nativeImageAtIndex
+
+    // If frame0 is displayed for the first time, startAnimation() has to request decoding frame1
+    // asynchronously. A flicker will occur if we request decoding frame0 also asynchronously.
+    if (UNLIKELY(options.decodingMode() == DecodingMode::Asynchronous && isAnimated() && !hasEverAnimated()))
```
I don't think the `UNLIKELY` will do anything useful.

shallawa replied:
`UNLIKELY` was removed.

smfr reviewed (approved):
A test would be good.

shallawa committed:
REGRESSION(/276827@main): Large animated images are always decoded asynchronously
https://bugs.webkit.org/show_bug.cgi?id=278593
rdar://134597661

Reviewed by Simon Fraser.

The large animated images should be decoded asynchronously if the setting
AnimatedImageAsyncDecodingEnabled is true. The only exception is the first frame
has to be decoded synchronously only for the first time.

* LayoutTests/fast/images/async-image-large-animated-disabled-expected.txt: Added.
* LayoutTests/fast/images/async-image-large-animated-disabled.html: Added.
* LayoutTests/fast/images/resources/animated-red-green-blue-400x400.gif: Added.
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::startAnimation):
(WebCore::BitmapImageSource::imageFrameDecodeAtIndexHasFinished):
(WebCore::BitmapImageSource::cacheNativeImageAtIndex):
(WebCore::BitmapImageSource::currentNativeImageForDrawing):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):

Canonical link: https://commits.webkit.org/283753@main
